gateway/api/views/jobs.py

```diff
@@ -249,6 +249,73 @@ def result(self, request, pk=None):  # pylint: disable=invalid-name,unused-argum
             serializer = self.get_serializer(job)
         return Response(serializer.data)
 
+    @action(methods=["POST"], detail=True)
+    def sub_status(self, request, pk=None):
+        """Update the sub status of a job."""
+        tracer = trace.get_tracer("gateway.tracer")
+        ctx = TraceContextTextMapPropagator().extract(carrier=request.headers)
+        with tracer.start_as_current_span("gateway.job.sub_status", context=ctx):
+            author = self.request.user
+            sub_status = self.request.data.get("sub_status")
+            if sub_status is None or sub_status not in Job.RUNNING_SUB_STATUSES:
+                return Response(
+                    {"message": "'sub_status' not provided or is not valid"},
+                    status=status.HTTP_400_BAD_REQUEST,
+                )
+
+            def set_sub_status():
+                job = self.jobs_repository.get_job_by_id(pk)
+
+                # If we import this with the regular imports,
+                # update jobs statuses test command fail.
+                # it should be something related with python import order.
+                from api.management.commands.update_jobs_statuses import (  # pylint: disable=import-outside-toplevel
+                    update_job_status,
+                )
+
+                update_job_status(job)
+
+                if job is None:
+                    return Response(
+                        {"message": f"Job [{pk}] not found"},
+                        status=status.HTTP_404_NOT_FOUND,
+                    )
+                can_update_sub_status = JobAccessPolocies.can_update_sub_status(
+                    author, job
+                )
+                if not can_update_sub_status:
+                    return Response(
+                        {"message": f"Job [{job.id}] not found"},
+                        status=status.HTTP_404_NOT_FOUND,
+                    )
+
+                if job.status != Job.RUNNING:
+                    logger.warning(
+                        "'sub_status' cannot change because the job"
+                        " [%s] current status is not Running",
+                        job.id,
+                    )
+                    return Response(
+                        {
+                            "message": "Cannot update 'sub_status' when is not"
+                            f" in RUNNING status. (Currently {job.status})"
+                        },
+                        status=status.HTTP_403_FORBIDDEN,
+                    )
+
+                job.sub_status = sub_status
+                job.save()
```

Comment:
I would like to propose another approach for this `save`. It could happen that as we are doing `update_job_status` to ensure that the Job is updated, since we retrieve the information from the DB the `status` could change also. So I would like to propose move these lines:
```
job.sub_status = sub_status
job.save()
```

To a new method in `JobRepository` called `jobs_repository.update_job_statuses` that will do something like:
```
updated = Job.objects.filter(id=id, status=Job.RUNNING).update(sub_status=sub_status)
```

This way we can ensure that the update happens at the moment the Job was in a Running state and we could check with `updated != 1` if we had success or not updating the object.

WDYT @korgan00 ?

Reply:
I'm ok with everything but, what happens if we are not succeed? In my opinion, we checked everything before and if we fail is because of a race condition so, we can safely say that is not an error.

Reply:
I'm ok with everything but, what happens if we are not succeed? In my opinion, we checked everything before and if we fail is because of a race condition so, we can safely say that is not an error.